npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -23,38 +24,49 @@ func (iMgr *IPSetManager) exists(name string) bool {
 	return ok
 }
 
-func NewIPSetManager() IPSetManager {
-	return IPSetManager{
-		setMap:      make(map[string]*IPSet),
-		dirtyCaches: make(map[string]struct{}),
+func NewIPSetManager() *IPSetManager {
+	iMgr := IPSetManager{
+		setMap: make(map[string]*IPSet),
 	}
+	iMgr.resetCaches()
+	return &iMgr
+}
+
+func (iMgr *IPSetManager) resetCaches() {
+	iMgr.dirtyCaches = make(map[string]struct{})
+	iMgr.deleteCache = make(map[string]string)
+}
+
+func (iMgr *IPSetManager) updateMapsOnDelete(set *IPSet) {
+	iMgr.deleteCache[set.Name] = set.HashedName
+	delete(iMgr.dirtyCaches, set.Name)
+	delete(iMgr.setMap, set.Name)
 }
 
 func (iMgr *IPSetManager) updateDirtyCache(setName string) {
 	set, exists := iMgr.setMap[setName] // check if the Set exists
-	if !exists {
+	if !exists {                        // TODO (hunter) shouldn't this never happen?
 		return
 	}
 
 	// If set is not referenced in netpol then ignore the update
-	if len(set.NetPolReference) == 0 && len(set.SelectorReference) == 0 {
-		return
-	}
+	// TODO commenting for testing
+	// TODO need to delete the set in the kernel once the references reach 0
+	// if len(set.NetPolReference) == 0 && len(set.SelectorReference) == 0 {
+	// 	return
+	// }
 
 	iMgr.dirtyCaches[set.Name] = struct{}{}
+	delete(iMgr.deleteCache, set.Name)
 	if set.Kind == ListSet {
 		// TODO check if we will need to add all the member ipsets
 		// also to the dirty cache list
 		for _, member := range set.MemberIPSets {
-			iMgr.dirtyCaches[member.Name] = struct{}{}
+			iMgr.updateDirtyCache(member.Name)
```

Comment:
if we keep the delete cache, perhaps we should keep this change since updateDirtyCache() both adds the set to the dirtyCache and deletes the set from deleteCache